### PR TITLE
Feature/trial api

### LINKED ISSUE #307: ModelCheckpoint: saving initial_epoch, early stopping state, reduced lr, and training history
So that we can resume training seamlessly from the epoch we KeyboardInterrupted at :-)

Happy to put in a PR for this. Any pointers on where to start?

### LINKED ISSUE #312: Make model class atomic
Further to issue #307 the Model class should be atomic, with everything being given in the constructor (except number of epochs to train for). Will need a good API for this as there will be many params to fill

### LINKED ISSUE #314: Create state key class and incorporate seed state into trial object constructor
This will allow for values to be seeded from state instead as well as being given in constructors, should enable grid search etc.

### LINKED ISSUE #324: Model resuming
Further to the discussion in #307, it would be nice for the resuming from checkpoint functionality to include some way of 'replaying' a model fit from the history. Since the model resuming is optional in the new API (can either resume or just load the model weights), I propose the following:

1. Change the run call to take 'max_epochs' instead of 'epochs', this should make no difference to those not resuming (or minor difference, could potentially offer both but it might be confusing)
2. Ammend the tqdm callback to print a kind of summary of the epochs until now if the initial epoch is not zero (i.e. when resuming). This might require storing more than just final metrics in history (although it would be nice to avoid storing running metrics in history, we could just take the value 10 times an epoch or something?)

The result of these two changes would be that: when resuming a model, the tqdm callback will 'catch you up' and you can have it simply 'replay' a fit call by loading in a model. An alternate option, would be to save all of the metrics, and have the fit call perform a callback run with the metrics. However, this seems a bit overkill as we would also need any other values from state that the callback uses. @kl2792 what are your thoughts on these? anything you would add?

### LINKED ISSUE #329: Inject sampler into state instead of callback list
We should inject the sampler into state[torchbearer.SAMPLER] instead of into the callback list. 
We can then call state[torchbearer.SAMPLER].sample() to get data instead of having it appear after callback_list.on_sample().

### LINKED ISSUE #334: Make callbacks saveable
Should have a state dict method which gets bundled with trial.state_dict

### PR DESCRIPTION
**Pull request by MattPainter01**
Closes #307, closes #312, closes #314, closes #324, closes #329, closes #334

### CODE REVIEW

File: torchbearer/callbacks/decorators.py [hunk 1]
```diff
@@ -249,13 +237,33 @@ def add_to_loss(func):
...
         state[torchbearer.LOSS] += func(state)
 
-    new_callback = Callback()
-    new_callback.on_criterion = add_to_loss_func
-    new_callback.on_criterion_validation = add_to_loss_func
-    return new_callback
+    return add_to_loss_callback
+
+
+def from_state(key_list):
```
Comment (ethanwharris) [inline torchbearer/callbacks/decorators.py]: Should this be removed?

File: torchbearer/state.py [hunk 2]
```diff
@@ -33,18 +85,33 @@ def state_key(key):
...
 LOSS = state_key('loss')
 FINAL_PREDICTIONS = state_key('final_predictions')
 BATCH = state_key('t')
 TIMINGS = state_key('timings')
 CALLBACK_LIST = state_key('callback_list')
+HISTORY = state_key('history')
+BACKWARD_ARGS = state_key('backward_args')
+
+# Legacy
+VALIDATION_ITERATOR = 'validation_iterator'
```
Comment (ethanwharris) [inline torchbearer/state.py]: Will these be set in the Model?
